### PR TITLE
Add basic multitouch support for osu!

### PR DESCRIPTION
**Pull request by o-dasher**
https://user-images.githubusercontent.com/88356162/186967880-0413298b-fd41-48b3-afcf-113e8fff3dc8.mp4

This is a simple implementation of osu! ruleset touch support, to make this implementation i followed the specifications at #18882, this could still be improved by considering the cursor touch to be the cursor closer to the current hittable object at the playfield, but this work is probably way too big for a single pull request.

### CODE REVIEW

File: osu.Game.Rulesets.Osu/OsuInputManager.cs [hunk 1]
```diff
@@ -43,17 +53,52 @@ protected override bool Handle(UIEvent e)
...
+
         protected override bool HandleMouseTouchStateChange(TouchStateChangeEvent e)
         {
-            if (!AllowUserCursorMovement)
+            var source = e.Touch.Source;
+
+            int touchNumber = numberedTouchSources[source];
+
+            if (touchNumber > simultaneous_touches_limit)
+                return false;
```
Comment (frenzibyte) [inline osu.Game.Rulesets.Osu/OsuInputManager.cs:69]: `TouchSource` is made in a way that allows mapping sources to integers without the need of any extra object or explicit dictionaries.
```suggestion
            var source = e.Touch.Source;
            if (source - TouchSource.Touch1 >= simultaneous_touches_limit)
                return false;
```

File: osu.Game.Rulesets.Osu/OsuInputManager.cs [hunk 2]
```diff
@@ -43,17 +53,52 @@ protected override bool Handle(UIEvent e)
             return base.Handle(e);
         }
 
+        private OsuAction getActionForTouchSource(TouchSource source)
+        {
+            int sourceNumber = numberedTouchSources[source];
+            return sourceNumber % 2 == 0 ? OsuAction.RightButton : OsuAction.LeftButton;
+        }
```
Comment (frenzibyte) [inline osu.Game.Rulesets.Osu/OsuInputManager.cs:60]: Same as above, `numberedTouchSources` is not required.
```suggestion
        private OsuAction getActionForTouchSource(TouchSource source) => (int)source % 2 == 0 ? OsuAction.RightButton : OsuAction.LeftButton;
```

File: osu.Game.Rulesets.Osu/OsuInputManager.cs [hunk 3]
```diff
@@ -43,17 +53,52 @@ protected override bool Handle(UIEvent e)
...
+
+            bool isCursorTouch = source == cursorTouch;
+            bool isTapTouch = !isCursorTouch;
+
+            foreach (var action in newInactiveTapActions)
+                KeyBindingContainer.TriggerReleased(action);
+
+            if (isTapTouch)
+            {
+                var action = getActionForTouchSource(source);
```
Comment (frenzibyte) [inline osu.Game.Rulesets.Osu/OsuInputManager.cs:89]: For consistency with this code, the action should be determined by the index of the touch in the active tap touches.
```suggestion
                int tapNumber = activeTapTouches.TakeWhile(t => t != source).Count();
                var action = tapNumber % 2 == 1 ? OsuAction.RightButton : OsuAction.LeftButton;

```

File: osu.Game.Rulesets.Osu/OsuInputManager.cs [hunk 4]
```diff
@@ -43,17 +53,52 @@ protected override bool Handle(UIEvent e)
...
+
+            var activeSources = CurrentState.Touch.ActiveSources;
+
+            var cursorTouch = activeSources.Any() ? activeSources.First() : source;
+
+            var activeTapTouches = AllowUserCursorMovement ? activeSources.Skip(1) : activeSources;
+            var limitedActiveTapTouches = activeTapTouches.Take(tap_touches_limit);
+
+            var newActiveTapActions = limitedActiveTapTouches.Select(getActionForTouchSource);
+            var newInactiveTapActions = PressedActions.Where(a => !newActiveTapActions.Contains(a)).ToList();
```
Comment (frenzibyte) [inline osu.Game.Rulesets.Osu/OsuInputManager.cs:79]: As mentioned above, this should be saved in private fields rather than reallocating every time a touch changes state.

File: osu.Game.Rulesets.Osu/OsuInputManager.cs [hunk 5]
```diff
@@ -15,6 +17,9 @@ namespace osu.Game.Rulesets.Osu
 {
     public class OsuInputManager : RulesetInputManager<OsuAction>
     {
+        private const int tap_touches_limit = 2;
+        private const int simultaneous_touches_limit = tap_touches_limit + 1;
```
Comment (frenzibyte) [inline osu.Game.Rulesets.Osu/OsuInputManager.cs:21]: Please add xmldoc explaining the difference between "tap" and "simultaneous".

**Review by frenzibyte — CHANGES_REQUESTED** [review]
Thanks for the contribution. I haven't tested this in mobile yet, but code-wise this needs a rework for the following reasons:
 - Key-binding actions should never be triggered inside the input propagation flow, let alone in the `HandleMouseTouchStateChange` path which is meant to propagate mouse input from touch, not handling touch.
 - There are multiple LINQ operations/allocations occurring every time a touch has changed state, this could get quite bad performance-wise.

Take a look at how touch support is added in osu!mania for inspiration, it's all isolated in a `Component` that handles touch events properly.

In addition, tests are immensely required to understand the implications of the logic you currently have here, as it doesn't look straightforward enough to not have one, and cleaning it up may potentially result in breaking actual behaviour (take a look at `OsuManualInputManagerTestScene` for sending input in a test scene and `CreateBeatmap` for defining a beatmap that's optimal for testing).

**Comment by o-dasher** [comment]
> Thanks for the contribution. I haven't tested this in mobile yet, but code-wise this needs a rework for the following reasons:
>  - Key-binding actions should never be triggered inside the input propagation flow, let alone in the `HandleMouseTouchStateChange` path which is meant to propagate mouse input from touch, not handling touch.
>  - There are multiple LINQ operations/allocations occurring every time a touch has changed state, this could get quite bad performance-wise.
> 
> Take a look at how touch support is added in osu!mania for inspiration, it's all isolated in a `Component` that handles touch events properly.
> 
> In addition, tests are immensely required to understand the implications of the logic you currently have here, as it doesn't look straightforward enough to not have one, and cleaning it up may potentially result in breaking actual behaviour (take a look at `OsuManualInputManagerTestScene` for sending input in a test scene and `CreateBeatmap` for defining a beatmap that's optimal for testing).

I think i've fixed the performance concerns, i will take a look at the mania touch implementation later, thanks.

**Comment by o-dasher** [comment]
Reading through the failed tests, i don't think these are related to this branch changes, so maybe it would be necessary to merge master branch onto this, but overall i think everything seems fine to me, after a few refactors and writing the tests and re-reading the code again and again it's logic seems to be pretty straightforward.

**Comment by smoogipoo** [comment]
I'm not really a fan of the cross communication between `OsuInputManager` and `OsuDrawableTouchInputHandler`. Since the latter is a child of the former, can't everything be implemented inside `OsuDrawableTouchHandler` itself?

**Comment by o-dasher** [comment]
> I'm not really a fan of the cross communication between `OsuInputManager` and `OsuDrawableTouchInputHandler`. Since the latter is a child of the former, can't everything be implemented inside `OsuDrawableTouchHandler` itself?

I don't think this cross communication is avoidable, because we can't make a touch be inactive through another drawable, which is the reason to the `DragMode` exist (Disable cursor touch and make it drag only so you can click with other tap finger), and it would be inconsistent even if it was possible because of other input managers children.

**Comment by peppy** [comment]
I've tidied things up for you, but there's some serious issues:

- I have no idea what the test scene is actually testing.
- It's definitely not testing enough. Reading your code I can see scenarios where it will completely fall over. Touches are not guaranteed to be in order, yet you're only reading the first x touches in the enum (the `AllowedTouches`). You will need to handle this dynamically, tracking the active touches and basing it on the number of touches down, not the enum.

**Review by peppy — CHANGES_REQUESTED** [review]
Needs better explained (and more) test coverage.

**Comment by o-dasher** [comment]
> - Touches are not guaranteed to be in order, yet you're only reading the first x touches in the enum (the `AllowedTouches`). You will need to handle this dynamically, tracking the active touches and basing it on the number of touches down, not the enum.
If you mean the "drag mode", now this part should be fixed.

**Comment by o-dasher** [comment]
> I've tidied things up for you, but there's some serious issues:
> 
> - I have no idea what the test scene is actually testing.
>   - There's no way to visualise what is going on.
>   - It's talking about dragging but the cursor isn't moving.
>   - It's not playing to the gameplay but using a player test.
>   - It's not testing any of the edge cases that would need testing.
> - It's definitely not testing enough. Reading your code I can see scenarios where it will completely fall over. Touches are not guaranteed to be in order, yet you're only reading the first x touches in the enum (the `AllowedTouches`). You will need to handle this dynamically, tracking the active touches and basing it on the number of touches down, not the enum.
I think i've addressed all these issues, including the test coverage, i think it is testing the breaking cases you had noticed now.

**Comment by o-dasher** [comment]
I think i've found a unwanted behaviour this being that when you click with three fingers it disables the left button (this because the first touch mouse input is disabled so we can stream properly), so you have to click once again with the third finger for it to activate the left button.

**Comment by o-dasher** [comment]
Autopilot touch support seems to be messed after these changes, i will try looking at it.

**Comment by o-dasher** [comment]
Ok now this is in a state that i am happy with.

**Comment by o-dasher** [comment]
It seems that the code quality check fails aren't my fault so i will be ignoring these then.

File: osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+using osu.Framework.Input;
+using osu.Framework.Input.Bindings;
+using osu.Framework.Input.Events;
+
+namespace osu.Game.Rulesets.Osu.UI
+{
+    public class OsuTouchInputMapper : Drawable
+    {
+        /// <summary>
+        /// The default cursor touch it may fail <see cref="IsCursorTouch(TouchSource)"/> on cases where there aren't any cursor touches, such as when autopilot is on.
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs:17]: I have no idea what this is trying to say. It feels like three sentences mashed into one, all touching on different points.

File: osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        /// A dictionary that maps <see cref="TouchSource"/> into a respective <see cref="OsuAction"/>.
+        /// </summary>
+        private readonly Dictionary<TouchSource, OsuAction> tapTouchActions = Enum.GetValues(typeof(TouchSource))
+                                                                                  .Cast<TouchSource>()
+                                                                                  .ToDictionary(source => source, source => (source - TouchSource.Touch1) % 2 == 0 ? OsuAction.LeftButton : OsuAction.RightButton);
+
+        /// <summary>
+        /// Tracks all active tap <see cref="TouchSource"/>s that can be mapped into a given <see cref="OsuAction"/> through the <see cref="tapTouchActions"/>.
+        /// </summary>
+        private readonly HashSet<TouchSource> activeTapTouches = new HashSet<TouchSource>();
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs:51]: Why is this required when `tapTouchActions` already exists? Seems like it's storing the same thing.

**Comment by peppy** [comment]
I was going to review this but can't bring myself to get through it. The logic is too confusing for my brain, sorry.

Try and refactor to remove `HandleTouchTapOnlyMapping` because this crumbles everything else in the PR. It's not feasible for this to be understandable by the average developer.

**Comment by o-dasher** [comment]
> I was going to review this but can't bring myself to get through it. The logic is too confusing for my brain, sorry.
> 
> Try and refactor to remove `HandleTouchTapOnlyMapping` because this crumbles everything else in the PR. It's not feasible for this to be understandable by the average developer.

I don't think that is possible to remove this, but i agree that it is confusing indeed, why this exists is too make the action of the first finger be disabled so the tap fingers fully take place of all the tapping done.
maybe comments explaining what everything on that method is doing should be helpful?

File: osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs (hunk 7 above)
Comment (o-dasher) [reply osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs:51]: It is required because these are the tap touch sources that were handled by the `OnTouchDown` so they can be removed and their assigned action be released later.
i though on just making a count of sctive tap touches and incrementing when a touch happens and getting the current count `OnTouchUp` and releasing the touch for the index of said count, but since tap ordering isn't guaranteed this does not work, so i believe this is a must.

File: osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs (hunk 6 above)
Comment (o-dasher) [reply osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs:17]: I will try to document it better when i get home, sorry.

File: osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs (hunk 7 above)
Comment (o-dasher) [reply osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs:51]: Also these active tap touches track how many tap touches are currently active so we don't accept tap input for too many fingers.

File: osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs (hunk 6 above)
Comment (o-dasher) [reply osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs]: I believe i explained better what the `DEFAULT_CURSOR_TOUCH` means on my recent commits.

**Comment by o-dasher** [comment]
> I was going to review this but can't bring myself to get through it. The logic is too confusing for my brain, sorry.
> 
> Try and refactor to remove `HandleTouchTapOnlyMapping` because this crumbles everything else in the PR. It's not feasible for this to be understandable by the average developer.

I renamed it to `BlockTouchCursorAction` which for me seems to be more readable and made comments explaining on what it is actually doing.

**Comment by o-dasher** [comment]
Hello, since releasing to ios was added to the roadmap for this month, it would be a nice addition to have atleast this base touchscreen support.
I am just asking for a review, because in a few months i may barely have any spare time to work on this anymore.
I will likely be changing to a full time school, but i am glad to help while i can with ease.

**Comment by peppy** [comment]
I have hope of either merging this or some alternative to this. It would be great if you could fix the conflicts and do a self-check of the code in preparation for that.

**Comment by peppy** [comment]
I started to fix this up but lost hope halfway through. I'll open a PR with my own version of this as I think it's the path of least resistance.

**Comment by o-dasher** [comment]
Understandable, i am sorry for the commits spam, is just that i had tried many ways to solve multiple problems when trying to get the implementation right.

**Comment by peppy** [comment]
The commit count is the least of the issue for me. The larger part is that the structure is not contained well enough. I'd hope to see very few changes to the `OsuInputManager` class, with all the logic being very localised in the touch handler instead.

**Comment by o-dasher** [comment]
Would you be fine if i open another pull request and try to do this instead. Or would you prefer to do it yourself?

**Comment by peppy** [comment]
I won't stop you, and while I do have relative high priority on this I do have a few higher prio things to visit first. You're more than welcome to continue work and see what you come up with. At very least I'll take inspiration from it (and potentially reuse some tests depending on their quality).

**Comment by peppy** [comment]
@o-dasher I'll likely revisit this this week. Have you made any progress?

**Comment by o-dasher** [comment]
> @o-dasher I'll likely revisit this this week. Have you made any progress?

I was going to get back to work on this last week although my keyboard broke, so i can't do nothing for now.

**Comment by peppy** [comment]
I've tested this PR as is on a device and it behaves completely wrong, so I'll likely be looking at things from scratch. Here's what I was seeing in some brief testing:

When attempting to use one finger to track and two more to tap:

https://user-images.githubusercontent.com/191335/212264004-68b5c15d-fc21-4e40-9eb8-36e24c475a3a.MP4


- Works close to expectations, but buttons can get stuck in an incorrect state at times, for instance B1 generally stays lit at all times even though I'm holding nothing.

**Comment by o-dasher** [comment]
Oh, ok then. Starting off from scratch may be better.
These errors where likely due to lack of test cases, i wasn't able to compile to android to see how stuff behaves on practice so this may had lead me to that.
I am sorry for that

**Comment by peppy** [comment]
Fair enough. I've been able to reuse some of your code so not all is lost (at very least I used i the tests as a starting point for a new attempt).
